File: tree/trie/208.Implement_Trie(PrefixTree).js
/*
208.
A trie (pronounced as "try") or prefix tree is a tree data structure used to efficiently store and retrieve keys in a dataset of strings. 
There are various applications of this data structure, such as autocomplete and spellchecker.
Implement the Trie class:
Trie() Initializes the trie object.
void insert(String word) Inserts the string word into the trie.
boolean search(String word) Returns true if the string word is in the trie (i.e., was inserted before), and false otherwise.
boolean startsWith(String prefix) Returns true if there is a previously inserted string word that has the prefix prefix, and false otherwise.
*/
class TrieNode {
  constructor(val = null) {
    this.val = val;
    this.children = []; //store nodes
    this.isEnd = false;
  }
}

class Trie {
  constructor() {
    //create empty new root first
    this.root = new TrieNode(-1);
  }

  insert(word) {
    let node = this.root;
    for (let c of word) {
      if (node.children[c] == null) node.children[c] = new TrieNode(c);
      node = node.children[c];
    }
    node.isEnd = true;
  }

  search(word) {
    let node = this.root;
    for (let c of word) {
      if (node.children[c] == null) return false;
      node = node.children[c];
    }
    return node.isEnd;
  }

  startsWith(prefix) {
    let node = this.root;
    for (let c of prefix) {
      if (node.children[c] == null) return false;
      node = node.children[c];
    }
    return true;
  }
}

/**
 * Your Trie object will be instantiated and called as such:
 * var obj = new Trie()
 * obj.insert(word)
 * var param_2 = obj.search(word)
 * var param_3 = obj.startsWith(prefix)
 */

1032. Stream of Characters
class TrieNode:
    def __init__(self, val=None):
        self.val = val
        self.children = [None] * 26
        self.isEnd = False
        
class StreamChecker:
    def __init__(self, words: List[str]):
        self.root = TrieNode()
        self.history = []
        for word in words:
            #store word from last char
            cur = self.root
            for i in range(len(word)-1, -1, -1):
                if not cur.children[ord(word[i]) - ord("a")]:
                    cur.children[ord(word[i]) - ord("a")] = TrieNode(word[i])
                cur = cur.children[ord(word[i]) - ord("a")] 
            cur.isEnd = True
        
    def query(self, letter: str) -> bool:
        self.history.append(letter)
        cur = self.root
        #each time query one time
        #start from newest to oldest see if can find whole word in words
        index = len(self.history) - 1
        while index >= 0 and cur.children[ord(self.history[index]) - ord("a")]:
            cur = cur.children[ord(self.history[index]) - ord("a")] 
            if cur and cur.isEnd: return True
            index -= 1
        return False
            



# Your StreamChecker object will be instantiated and called as such:
# obj = StreamChecker(words)
# param_1 = obj.query(letter)

File: Tree/208. Implement Trie (Prefix Tree).py
'''
TOPIC:dictionary tree, use For search word if exist
STEP:
1.create TrieNode include isEnd, children 26List of nodes, charvar
2.search/insert
'''
class TrieNode:
    def __init__(self, val=None):
        self.val = val
        self.isEnd = False
        self.children = [None] * 26 #each node has 26 children node. default is None
    
class Trie:
    def __init__(self):
        #start from empty root
        self.root = TrieNode()
        
    def insert(self, word: str) -> None:
        #each time insert or search, start from root
        cur = self.root
        for c in word:
            if cur.children[ord(c) - ord('a')] is None: #dictionary here only have lowercase
                #not exist yet, need create new trinode
                cur.children[ord(c) - ord('a')] = TrieNode(c)
            cur = cur.children[ord(c) - ord('a')]
        cur.isEnd = True
                  

    def search(self, word: str) -> bool:
        cur = self.root
        for c in word:
            if cur.children[ord(c) - ord('a')] is None:
                return False
            else:
                cur = cur.children[ord(c) - ord('a')]
        return cur.isEnd == True
        

    def startsWith(self, prefix: str) -> bool: # have apple, now word is app -> true
        cur = self.root
        for c in prefix:
            if cur.children[ord(c) - ord('a')] is None:
                return False
            else:
                cur = cur.children[ord(c) - ord('a')]
        return True
        


# Your Trie object will be instantiated and called as such:
# obj = Trie()
# obj.insert(word)
# param_2 = obj.search(word)
# param_3 = obj.startsWith(prefix)